docarray/index/backends/redis.py

```diff
@@ -0,0 +1,609 @@
+        :return: Filter results.
+        """
+        results = []
+        for query in filter_queries:
+            results.append(self._filter(filter_query=query, limit=limit))
+        return results
+
+    def _filter_by_parent_id(self, id: str) -> Optional[List[str]]:
+        """Filter the ids of the subindex documents given id of root document.
+
+        :param id: the root document id to filter by
+        :return: a list of ids of the subindex documents
+        """
+        docs = self._filter(filter_query=f'@parent_id:{{{id}}}', limit=self.num_docs())
+        return [doc['id'] for doc in docs]
+
+    def _text_search(
+        self, query: str, limit: int, search_field: str = ''
+    ) -> _FindResult:
+        """
+        Conducts a text-based search on the index.
+
+        :param query: The query to search.
+        :param limit: The maximum number of results to return.
+        :param search_field: The field to search the query.
+        :return: Search results.
+        """
+        query_str = '|'.join(query.split(' '))
+        q = (
+            Query(f'@{search_field}:{query_str}')
+            .scorer(self._text_scorer)
+            .with_scores()
+            .paging(0, limit)
+        )
+
+        results = self._client.ft(index_name=self._index_name).search(q).docs
+
+        scores: NdArray = NdArray._docarray_from_native(
+            np.array([document['score'] for document in results])
+        )
+
+        docs = [json.loads(doc.json) for doc in results]
+
+        return _FindResult(documents=docs, scores=scores)
+
+    def _text_search_batched(
+        self, queries: Sequence[str], limit: int, search_field: str = ''
+    ) -> _FindResultBatched:
+        """
+        Conducts a batched text-based search on the index.
+
+        :param queries: The queries to search.
+        :param limit: The maximum number of results to return for each query.
+        :param search_field: The field to search the queries.
+        :return: Search results.
+        """
+        docs, scores = [], []
+        for query in queries:
+            results = self._text_search(
+                query=query, search_field=search_field, limit=limit
+            )
+            docs.append(results.documents)
+            scores.append(results.scores)
+
+        return _FindResultBatched(documents=docs, scores=scores)
+
+    def __contains__(self, item: BaseDoc) -> bool:
+        """
+        Checks if a given document exists in the index.
+
+        :param item: The document to check.
+            It must be an instance of BaseDoc or its subclass.
+        :return: True if the document exists in the index, False otherwise.
+        """
+        if safe_issubclass(type(item), BaseDoc):
+            return self._doc_exists(item.id)
+        else:
+            raise TypeError(
+                f"item must be an instance of BaseDoc or its subclass, not '{type(item).__name__}'"
+            )
```

Comment:
Maybe worthy of a separete PR, but I think we should move `__contains__` to the abstract class and tell backends to implement `_doc_exitsts()`.

Reply:
https://github.com/docarray/docarray/issues/1686